### PR TITLE
backend: dwarfdbginf: refactor DWARF abbreviations

### PR DESCRIPTION
**Pull request by ljmf00**
Signed-off-by: Luís Ferreira <contact@lsferreira.net>

---

Motivation for this change:
- All abbreviations are LEB128 encoded and a lot of code uses `byte` writes, which is technically right, because, before a certain range (> `0x80`), LEB128 encoding doesn't matter, but in situations like https://github.com/dlang/dmd/pull/13202 , people (like me) struggle at finding why such simple addition doesn't work as expected. Therefore this increases a lot on readability and simplifies the logic.
- Compile-time LEB128 encoding, increasing performance for bulk append.
- Modernize the code, but still maintain compatibility with C API.

### CODE REVIEW

**Comment by RazvanN7** [comment]
@ljmf00 are you still working on this?

**Comment by ljmf00** [comment]
Yes, I still have notes about this on my task list. I stopped this due to the fact that I have a lot of other changes on DWARF generator that will need refactoring for sure. I'm going to wait for merge of those changes and then proceed with the rest of the refactor.

**Comment by RazvanN7** [comment]
Can you point out what other PRs are stalling this? I can put them on the fast lane.

**Comment by ljmf00** [comment]
> Can you point out what other PRs are stalling this? I can put them on the fast lane.

I still have the following:
- https://github.com/dlang/dmd/pull/13274
- https://github.com/dlang/dmd/pull/13246

After I have the type naming correct (#13274), I need to do a similar change on top to https://github.com/dlang/dmd/pull/13254 for `char` and `wchar` types to make those types UTF encoded too. After that I think I can proceed to work on this.

**Comment by RazvanN7** [comment]
Ok, I merged one of those and the other one should get in soon. Ping me whenever you are stalled on something.

**Comment by ljmf00** [comment]
Weirdly, I can't reproduce dshell testsuite locally.

**Comment by thewilsonator** [comment]
This is green, is it ready for review?

**Comment by ljmf00** [comment]
> This is green, is it ready for review?

I still have a lot to change. This can be iteratively done tho. I'm thinking of adding more testing on DWARF before proceeding, but probably a half-backed refactor is not good to merge. The test suite I want to add is pretty much adding support for evaluating the debug info generated with `llvm-dwarfdump --verify`.

**Comment by ljmf00** [comment]
> This is green, is it ready for review?

I still have a lot to change. This can be iteratively done tho. I'm thinking of adding more testing on DWARF before proceeding, but probably a half-backed refactor is not good to merge. The test suite I want to add is pretty much adding support for evaluating the debug info generated with `llvm-dwarfdump --verify`.

**Comment by ljmf00** [comment]
Needs https://github.com/dlang/dmd/pull/13237 .

**Comment by thewilsonator** [comment]
Circular reference, stack overflow.

**Comment by ljmf00** [comment]
> Circular reference, stack overflow.

Oh, sorry 🙃 I meant #13754.

**Comment by ljmf00** [comment]
Finally ready for review :blush:

**Comment by WalterBright** [comment]
This is very hard to review because it tries to do way too much. The "Motivation" section lists 4 distinct refactorings. 4 PRs, one for each, would be much more reviewable. This one is nearly 500 lines long.

**Comment by ljmf00** [comment]
> This is very hard to review because it tries to do way too much. The "Motivation" section lists 4 distinct refactorings. 4 PRs, one for each, would be much more reviewable. This one is nearly 500 lines long.

I try to make refactors atomic, but this one I didn't see the need.

This PR pretty much removes manual LEB128 encoded arrays with `abuf.writeByte()` calls and turns them into compile-time LEB128 arrays writing it with a separate interface for DWARF Abbreviations. Everything listed in the motivation is tied up together. `__gshared` variables are removed because they don't make sense with that approach.

The hardest part I see in the review is the move of the `dwarf_abbrev_code` function, because `git` tries to be smart. I can split it up in a separate PR, if you prefer. The rest is just the same thing but done in the entire file, which is naturally huge. What I can do is do that changes incrementally, as they are independent of each other, although they are all the same refactor.

**Comment by ljmf00** [comment]
@WalterBright I just split this PR up into smaller pieces. I will create upcoming PRs to refactor the missing parts.

**Comment by ljmf00** [comment]
Ping @WalterBright or perhaps @RazvanN7 ca you review this?

File: src/dmd/backend/dwarfdbginf.d [hunk 1]
```diff
@@ -3194,11 +3162,9 @@ static if (1)
              */
             abbrev_table = AApair.create(debug_abbrev.buf.bufptr);
 
-        /* Write new entry into debug_abbrev.buf
```
Comment (thewilsonator) [inline src/dmd/backend/dwarfdbginf.d:3197]: Remove all changes below this to a separate PR

File: src/dmd/backend/dwarfdbginf.d [hunk 2]
```diff
@@ -2264,82 +2261,62 @@ static if (1)
...
 
-        static immutable ubyte[10] abbrevTypeBasic =
-        [
-            DW_TAG_base_type,       DW_CHILDREN_no,
-            DW_AT_name,             DW_FORM_string,
-            DW_AT_byte_size,        DW_FORM_data1,
-            DW_AT_encoding,         DW_FORM_data1,
-            0,                      0,
-        ];
-        static immutable ubyte[10] abbrevWchar =
```
Comment (thewilsonator) [inline src/dmd/backend/dwarfdbginf.d:2275]: was this unused?

File: src/dmd/backend/dwarfdbginf.d (hunk 1 above)
Comment (ljmf00) [reply src/dmd/backend/dwarfdbginf.d:3197]: Ah, thanks, forgot to split that.

File: src/dmd/backend/dwarfdbginf.d (hunk 2 above)
Comment (ljmf00) [reply src/dmd/backend/dwarfdbginf.d:2275]: Not unused, but duplicated with the above. I'm reusing the same abbreviation for basic types: `abbrevTypeBasic`.

Comment (thewilsonator) [reply src/dmd/backend/dwarfdbginf.d:2275]: Ah, OK

File: src/dmd/backend/dwarfdbginf.d (hunk 1 above)
Comment (ljmf00) [reply src/dmd/backend/dwarfdbginf.d]: Done

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**